src/adr/0043-avoid-elasticsearch-on-critical-reads.md

```diff
@@ -0,0 +1,161 @@
+# 43. Avoid ElasticSearch on critical reads
+
+Date: 2020-11-11
+
+## Status
+
+Accepted (lazy consensus).
+
+Scope: Distributed James
+
+## Context
+
+A user willing to use a webmail powered by the JMAP protocol will end up doing the following operations:
+ - `Mailbox/get` to retrieve the mailboxes. This call is resolved against metadata stored in Cassandra.
+ - `Email/query` to retrieve the list of emails. This call is nowadays resolved on ElasticSearch for Email search after
+ a right resolution pass against Cassandra.
+ - `Email/get` to retrieve various levels of details. Depending of requested properties, this is either
+ resolved on Cassandra alone or on ObjectStorage.
+
+So, ElasticSearch is queried on every JMAP interaction. Administrators thus need to enforce availability and good performance
+for this component.
+
+Relying on more software for every read also harms our resiliency as ElasticSearch outages have major impacts.
+
+Also we should mention our ElasticSearch implementation in Distributed James suffers the following flaws:
+ - Updates of flags lead to updates of the all Email object, leading to sparse segments
+ - We currently rely on scrolling for JMAP (in order to ensure messageId uniqueness in the response while respecting limit & position)
+ - We noticed some very slow traces against ElasticSearch, even for simple queries.
+
+Regarding Distributed James data-stores responsibilities:
+ - Cassandra is the source of truth for metadata, its storage needs to be adapted to known access patterns.
+ - ElasticSearch allows resolution of arbitrary queries, and performs full text search.
+
+## Decision
+
+Provide an optional view for most common `Email/query` requests both on Draft and RFC-8621 implementations.
+This includes filters and sorts on 'sentAt'.
+
+This view will be stored on Cassandra, and updated asynchronously via a MailboxListener.
+
+## Consequences
+
+A migration task will be provided for new adopters.
+
+Administrators would be offered a configuration option to turn this view on and off as needed.
+
+If enabled administrators would no longer need to ensure high availability and good performances for ElasticSearch.
```

Comment:
untrue, you can't assume people won't trigger complex queries

Reply:
> untrue, you can't assume people won't trigger complex queries

I agree with this statement.

However complex queries won't happen for carefully crafted "basic email display queries" our users are careful about using for their basic use cases <3

Reply:
What's the point having global queries if you don't expect to build valuable features based on that? Having only simple queries sounds like: we just want a simple webmail to me.

Reply:
We just want to have a simple that always work **.** If that webmail exposes some advanced features that have a lower availability, then it is nice too.

Reply:
I don't agree at all but it's not a problem because which webmail you want to offer to your customer is not a James matter (and this feature is optimal).
Could we just drop this sentence as it's not a consensus?

Reply:
No. That is one of the more important here.

It is opt-in: you do not necessarily have to consider es as non critical.

But if you want, you can.

What is the problem about alliwing diverging phylosophies?

Reply:
This sentence comes after a lot of assumptions: we know the usage pattern of users, we know the client they use, we don't care about QoS of services (ES in this case), etc.

The discussion we just had make these things apparent but the ADR does not AFAICT. 

I'm afraid people draw false conclusion by reading the ADR without the comments.

Reply:
Clarified: `given clients following well defined Email/query requests` + `of basic usages (mailbox content listing)`

Out of these assumptions, of course ElasticSearch is a critical dependency.